### DISCUSSION TITLE
Airflow + Okta  integration problem OAuth2.0

### DISCUSSION THREAD

**Discussion by aruiz-personal** [apache/airflow#17471] [Q&A]
If this is not the right place for this question, I apologize in advance.

I'm trying to authenticate to Airflow using Okta (our Idp) and OAuth2.0/OpenID. For this I used this as reference:
- https://tech.scribd.com/blog/2021/integrating-airflow-and-okta.html
- https://flask-appbuilder.readthedocs.io/en/latest/security.html

The problem is that after finishing all the configuration when I tried to log in using Okta I get the message "Invalid login. Please try again." 
![image](https://user-images.githubusercontent.com/13920987/128525271-74319429-a337-480f-a9e2-7d2f3e1a090c.png)
![image](https://user-images.githubusercontent.com/13920987/128525444-e97938c0-1900-4555-a10f-f74639f1b079.png)
![image](https://user-images.githubusercontent.com/13920987/128525569-15e5a30f-97ff-4d0a-a22f-940de05545a0.png)

The only logs I see coming from Flask i suppose are like:

172.17.0.1 - - [06/Aug/2021:14:05:01 +0000] "GET /static/pin_32.png HTTP/1.1" 304 0 "http://localhost:8080/login/?next=http%3A%2F%2Flocalhost%3A8080%2Fhome" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36"

172.17.0.1 - - [06/Aug/2021:14:05:26 +0000] "GET /login/okta?next=http://localhost:8080/home HTTP/1.1" 302 985 "http://localhost:8080/login/?next=http%3A%2F%2Flocalhost%3A8080%2Fhome" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36"

172.17.0.1 - - [06/Aug/2021:14:05:41 +0000] "GET /oauth-authorized/okta?code=V3YfnQL9IQOIwJLWt1KdALrtRRYM1xtZBDxNP9exlp4&state=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJuZXh0IjpbImh0dHA6Ly9sb2NhbGhvc3Q6ODA4MC9ob21lIl19.GbawEXaz1lpX1nOYuGyHHSstM9b-X36sghlhBoLuot0 HTTP/1.1" 302 221 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36"

172.17.0.1 - - [06/Aug/2021:14:05:41 +0000] "GET /login/ HTTP/1.1" 200 17174 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36"

I see nothing wrong there..

**Configuration steps:**

**1. Configuring Okta:**
![image](https://user-images.githubusercontent.com/13920987/128523704-80f2610a-f2c0-4c7e-a7eb-b334e79bd7ca.png)

Also the "Authorization server is the default one"

**2. Complete the webserver_config.py**

import os
from flask_appbuilder.security.manager import AUTH_OAUTH

AUTH_TYPE = AUTH_OAUTH
OAUTH_PROVIDERS = [
{'name': 'okta', 'icon': 'fa-circle-o',
        'token_key': 'access_token',
        'remote_app': {
            'client_id': '0oa1ceaw9cxc1rXhi5d7',
            'client_secret': 'MY CLIENT SECRET',
            'api_base_url': 'https://MY OKTA DOMAIN/oauth2/default/v1/',
            'client_kwargs': {
                'scope': 'openid profile email groups'
            },
            'access_token_url': 'https://MY OKTA DOMAIN/oauth2/default/v1/token',
            'authorize_url': 'https://MY OKTA DOMAIN/oauth2/default/v1/authorize',
    }
    }
]

**3. Create a Dockerfile:**
FROM apache/airflow
RUN pip install authlib\
	 && pip install flask-appbuilder==3.2.2\
   	 && pip install sqlalchemy==1.3.18\
COPY ./webserver_config.py /opt/airflow/webserver_config.py

**4. Build the image and run the airflow container:**
#docker build -t airflowcustom .
#docker run -d -p 8080:8080 --name airflow airflowcustom  webserver

I'm stuck here ... did anyone do this or experience a similar problem? 
Do you know what logs can I see and where? I don't see anything inside the airflow container.

**Comment by potiuk** [comment]
I think what might work for you (it worked for me last time when I had similar problems with a Google authentication) is to enable automated user creation. 

By default the oauth integration will not create new users in the DB - you have to have them in the db with the right email address and i think it is not obvious.

What I found necessary was two things:

1.Enable self-registration

https://github.com/apache/airflow/blob/ed1eac1457bdbab52e6ceb324a5a0bf4d8318a93/airflow/config_templates/default_webserver_config.py#L56

2. Setting the role for registered users - to one that is enabled ('Public' is disabled by default).

https://github.com/apache/airflow/blob/ed1eac1457bdbab52e6ceb324a5a0bf4d8318a93/airflow/config_templates/default_webserver_config.py#L70

Not sure if that works for you but if it does i think we need some update in our docs to point to those settings, as I think it is a bit 'buried'.

**Reply by potiuk** [reply]
Let me know if that helped :)

**Reply by aruiz-personal** [reply]
Hi @potiuk. Thanks for the reply.

I made the configuration you suggested, the problem that occurs in that case is another:

![image](https://user-images.githubusercontent.com/13920987/128732103-4d5088ac-7e02-4ab3-b6ca-5271937fef98.png)

And after put the user credentials in Okta Login we get this:

![image](https://user-images.githubusercontent.com/13920987/128732146-8c6aa04d-1451-4926-a243-913c32679f60.png)

I added the _docker logs airflow_ output here to.

With the options AUTH_USER_REGISTRATION = True AUTH_USER_REGISTRATION_ROLE = "Public" :
[logs.txt](https://github.com/apache/airflow/files/6955625/logs.txt)

Without the options, original config:
[logs_1.txt](https://github.com/apache/airflow/files/6955644/logs_1.txt)

**Reply by potiuk** [reply]
Try `AUTH_USER_REGISTRATION_ROLE` = Admin (That will make all your new users registered ad `Admin` though. What you likely have now, the user is created with `Public` role but the Public Role is not enabled so you get continuous error redirection.

Make sure that you know your Authentication/Authorisation setup you want to have - which roles should be disabled/enabled - which role should be assigned when users self-register.

**Reply by aruiz-personal** [reply]
Hi @potiuk, Thank u very much for the assistance. It was really helpful.

I also had to add a "airflow db init" command... For example the basic Dockerfile that worked for me during tests:
```
FROM apache/airflow
RUN pip  install authlib\
	 && pip install flask-appbuilder==3.2.2\
   	 && pip install sqlalchemy==1.3.18
COPY ./config_prod.py /opt/airflow/webserver_config.py
RUN airflow db init
```
I ll be writing a post with all the steps and configs and i'll share the link here when it's ready.

Bye.

**Comment by potiuk** [comment]
FYI. Your article might get quickly outdated at least for the installation part ;). 

Those all `RUN pip' won't be needed in Airflow 2.1.3 (release candidate will be out likely tomorrow). 
we've added authlib and FAB is "forced" to be 3.3.2 and above:

* https://github.com/apache/airflow/blob/0bd955831657a0ce628577983e5fb9ca16b7875c/setup.py#L273
* https://github.com/apache/airflow/blob/0bd955831657a0ce628577983e5fb9ca16b7875c/setup.cfg#L103

**Reply by potiuk** [reply]
Also - you should not run `airflow db init` in docker build phase. It should be run after that when you initialise your deployment - whether it is docker-compose or helm, it should be run when you initialize it. For test/non-production use you run the DB upgrade automatically when you start any container by setting appropriate env variable: https://airflow.apache.org/docs/docker-stack/entrypoint.html#upgrading-airflow-db - but this is not recommended for production.

**Reply by potiuk** [reply]
If you think about "serious" deployment - I recommend you switch to K8S + the official Helm Chart from the community - it is really good and stable and has a lot of deployment configuration solved for you (including running the DB upgrade when needed) https://airflow.apache.org/docs/helm-chart/stable/index.html

**Comment by aruiz-personal** [comment]
Hello again!! I will definitely take a look to all this information. We are actually doing this test as the first step of a migration plan to K8s! 

Allow me the audacity to take advantage of your knowledge one more time (I hope you don't mind). 
I added the following to the webserver_config.py:

AUTH_USER_REGISTRATION_ROLE = "Admin"
AUTH_ROLES_SYNC_AT_LOGIN = True
AUTH_ROLES_MAPPING = {
    "okta_users": ["User"],
    "okta_admin": ["Admin"],
}

However users are still assigned the "Admin" role no matter what. I was hoping that I could associate them with a different role according to Okta's group.

If you have documentation to share that would be great!!

Thank u.

**Reply by potiuk** [reply]
Hmm. I think most of the docs are in FAB documentation (https://flask-appbuilder.readthedocs.io/en/latest/) as we are using it under-the-hood (but let's say a lot of the "glue" between FAB and airflow is kinda not document beyond comments in the config). I think that part needs a bit overhaul when it comes to documentation (to be honest I was not too involved in that part before so my knowledge is more from my own trials and looking at the config/source code/guessing it a bit),

From what I understand Fab configuration the AUTH_USER_REGISTRATION_ROLE and AUTH_USER_REGISTRATION = True
are "overriding" any settings that are specific to authentication method - the AUTH_ROLES_MAPPING is only working for Okta now and when it works, you do not need to set AUTH_USER_REGISTRATION = True. 

But I have a bit limited knowledge here :).

**Reply by aruiz-personal** [reply]
Thank you very much for your assistance @potiuk!
